File: service_catalog/hooks/services.py
"""MkDocs hook that renders the digital service catalogue.

WHY THIS EXISTS
---------------
Editors of this site are librarians and research-support staff, not developers.
The catalogue must therefore be maintainable by dropping a single markdown file
into a folder — no central registry, no JSON index, no build step to remember.
This hook is what makes that possible: it turns the markdown files next to the
overview page into cards, filter dropdowns, clickable tags and modal windows.

HOW IT PLUGS IN
---------------
`mkdocs.yml` registers this file under `hooks:`. MkDocs then calls
:func:`on_page_markdown` for every page. Only pages containing the marker
``<!-- SERVICES_ARCHIVE -->`` are touched; the marker is replaced by the
generated HTML block. Everything else passes through unchanged.

INPUTS
    * `docs/<lang>/services/index.md` — the overview page carrying the marker.
    * `docs/<lang>/services/<service>.md` — one file per service, YAML frontmatter
      plus `##` sections (Access / Guides / About the service).

OUTPUT
    A single raw-HTML block injected into the overview page: filter form,
    card grid, one modal per service.

SIDE EFFECTS
    Reads files from disk and writes warnings to the MkDocs log. It never
    writes files; strict builds turn these warnings into build failures only
    for MkDocs' own warnings, not for ours.

ASSUMPTIONS
    * Service pages live in the SAME directory as the overview page.
    * Language is derived from the source path prefix (`sv/` → Swedish).
    * The generated markup is raw HTML, so styling and behaviour come from
      `docs/stylesheets/service_catalog/services.css` and
      `docs/javascripts/service_catalog/services.js`. The class names below are
      a contract with those two files — renaming one means renaming all three.

DEPENDENCIES
    markdown, PyYAML and mkdocs-material (for the inline icon SVGs).
"""

from __future__ import annotations

import html
import logging
import os
import re
from typing import Any, Iterable

import markdown as md_lib
import yaml

log = logging.getLogger("mkdocs.hooks.services")

# --------------------------------------------------------------------------- #
# Constants — no magic values further down.
# --------------------------------------------------------------------------- #

#: Placeholder in the overview page that gets replaced by the generated HTML.
MARKER = "<!-- SERVICES_ARCHIVE -->"

#: Rendered wherever an editor left a required field empty. The specification
#: requires missing information to be *visible* rather than silently omitted.
MISSING = "uppgift saknas"

#: Markdown extensions used when rendering the body sections of a service file.
MD_EXTENSIONS = ["admonition", "attr_list", "md_in_html", "tables", "pymdownx.details"]

#: Front matter keys that become tags, mapped to the CSS colour modifier used
#: for them (`.svc-tag--provider`, `.svc-tag--group`, ...). Only the first three
#: have a matching dropdown; `data` and `location` are display/search only.
TAG_GROUPS: tuple[tuple[str, str], ...] = (
    ("provider", "provider"),
    ("group", "group"),
    ("type", "type"),
    ("data", "data"),
    ("location", "location"),
)

#: Front matter keys a service file must define; a missing one is logged and
#: rendered as MISSING instead of breaking the build.
REQUIRED_FIELDS = ("name", "provider", "group", "summary")

#: Files in the services folder that are never treated as a service.
SKIPPED_FILENAMES = ("index.md",)
SKIPPED_PREFIX = "_"

#: Icon used when a service file does not name one.
DEFAULT_ICON = "material/apps"

#: Heading level that separates the expandable sections inside a service file.
SECTION_PREFIX = "## "

_FRONT_MATTER = re.compile(r"^---\s*\n(.*?)\n---\s*\n?", re.DOTALL)
_NON_SLUG_CHARS = re.compile(r"[^a-z0-9]+")


# --------------------------------------------------------------------------- #
# Small helpers
# --------------------------------------------------------------------------- #


def _markdown_renderer() -> md_lib.Markdown:
    """Create a fresh Markdown renderer.

    A new instance per section is deliberate: python-markdown instances keep
    state (footnotes, reference links) between conversions, which would leak
    content from one service card into the next.
    """
    return md_lib.Markdown(extensions=MD_EXTENSIONS)


def _slug(value: str) -> str:
    """Turn a human label into a stable identifier used in HTML and URLs.

    The same function is used for dropdown option values, card `data-*`
    attributes and tag values, which is what makes "click a tag" equal
    "select that dropdown option" on the client side.
    """
    return _NON_SLUG_CHARS.sub("-", value.lower()).strip("-")


def _as_list(value: Any) -> list[str]:
    """Normalise a front matter value to a list of non-empty strings.

    Editors may write either `data: Öppna data` or a YAML list; both must work.
    """
    if value is None or value == "":
        return []
    if isinstance(value, (list, tuple)):
        return [str(item) for item in value if str(item).strip()]
    return [str(value)]


def _split_sections(body: str) -> list[tuple[str, str]]:
    """Split a service file body into ``(heading, markdown)`` pairs on ``## ``.

    Content before the first `##` heading is intentionally dropped: the card
    summary comes from front matter, so stray prose above the first section
    would have no place to go in the modal.
    """
    sections: list[tuple[str, str]] = []
    current_title: str | None = None
    buffer: list[str] = []
    for line in body.splitlines():
        if line.startswith(SECTION_PREFIX):
            if current_title is not None:
                sections.append((current_title, "\n".join(buffer).strip()))
            current_title = line[len(SECTION_PREFIX):].strip()
            buffer = []
        elif current_title is not None:
            buffer.append(line)
    if current_title is not None:
        sections.append((current_title, "\n".join(buffer).strip()))
    return sections


# --------------------------------------------------------------------------- #
# Parsing service files
# --------------------------------------------------------------------------- #


def _build_tags(meta: dict[str, Any]) -> list[dict[str, str]]:
    """Build the tag list for one service from its front matter.

    Returns dicts of ``label`` (shown), ``kind`` (colour + which dropdown it
    drives) and ``value`` (slug compared against the dropdown value).
    """
    tags: list[dict[str, str]] = []
    for key, kind in TAG_GROUPS:
        for label in _as_list(meta.get(key)):
            tags.append({"label": label, "kind": kind, "value": _slug(label)})
    return tags


def _search_blob(service: dict[str, Any], tags: list[dict[str, str]]) -> str:
    """Pre-compute the lowercase text the free-text search matches against.

    Doing this at build time keeps the client-side search to a single
    `indexOf` per card, which is why filtering stays instant without a search
    library.
    """
    parts = [
        service["name"],
        service["provider"],
        service["group"],
        service["type"],
        service["summary"],
        service["access"],
    ] + [tag["label"] for tag in tags]
    return " ".join(parts).lower()


def _parse_service(path: str, docs_dir: str) -> dict[str, Any] | None:
    """Parse one service markdown file into the dict the renderers consume.

    Args:
        path: Absolute path to the service markdown file.
        docs_dir: MkDocs `docs_dir`, used to record a readable source path.

    Returns:
        A service dict, or None when the file has no/unreadable front matter.

    Side effects: reads the file and logs warnings for missing fields.
    """
    with open(path, encoding="utf-8") as handle:
        raw = handle.read()

    match = _FRONT_MATTER.match(raw)
    if not match:
        log.warning("services: %s has no front matter, skipping", path)
        return None

    try:
        meta = yaml.safe_load(match.group(1)) or {}
    except yaml.YAMLError as exc:
        log.warning("services: could not read front matter in %s (%s)", path, exc)
        return None

    for key in REQUIRED_FIELDS:
        if not meta.get(key):
            log.warning("services: %s is missing '%s' — rendered as '%s'", path, key, MISSING)

    body = raw[match.end():]
    filename = os.path.basename(path)

    service: dict[str, Any] = {
        "id": _slug(os.path.splitext(filename)[0]),
        "name": meta.get("name") or MISSING,
        "icon": meta.get("icon") or DEFAULT_ICON,
        "provider": meta.get("provider") or MISSING,
        "group": meta.get("group") or MISSING,
        "type": meta.get("type") or "",
        "summary": meta.get("summary") or MISSING,
        "access": meta.get("access") or MISSING,
        "link": meta.get("link") or "",
        "link_login": bool(meta.get("link_requires_login")),
        # Service pages sit next to the overview page, so a relative
        # "<name>/" link works in both languages and under a sub-path deploy.
        "page": re.sub(r"\.md$", "/", filename),
        "tags": _build_tags(meta),
        "sections": [
            {"title": title, "html": _markdown_renderer().convert(section_md)}
            for title, section_md in _split_sections(body)
        ],
        "source": os.path.relpath(path, docs_dir).replace(os.sep, "/"),
    }
    service["search"] = _search_blob(service, service["tags"])
    return service


def _collect(page: Any, config: Any) -> list[dict[str, Any]]:
    """Collect every service defined next to the overview page.

    Sorted by provider then name so the grid order is stable between builds
    (a changing order would produce noisy diffs in the generated site).
    """
    docs_dir = config["docs_dir"]
    services_dir = os.path.dirname(os.path.join(docs_dir, page.file.src_path))
    if not os.path.isdir(services_dir):
        return []

    services: list[dict[str, Any]] = []
    for filename in sorted(os.listdir(services_dir)):
        if not filename.endswith(".md"):
            continue
        if filename.startswith(SKIPPED_PREFIX) or filename in SKIPPED_FILENAMES:
            continue
        service = _parse_service(os.path.join(services_dir, filename), docs_dir)
        if service:
            services.append(service)
    services.sort(key=lambda s: (s["provider"].lower(), s["name"].lower()))
    return services


# --------------------------------------------------------------------------- #
# HTML rendering
# --------------------------------------------------------------------------- #


def _icon_html(icon: str) -> str:
    """Inline a Material icon SVG.

    The markup is emitted as raw HTML, so the usual `:material-icon:` shortcode
    is not processed here — the SVG has to be inlined directly.
    """
    try:
        import material

        base = os.path.join(os.path.dirname(material.__file__), "templates", ".icons")
        path = os.path.join(base, *f"{icon}.svg".split("/"))
        if os.path.isfile(path):
            with open(path, encoding="utf-8") as handle:
                return f'<span class="svc-icon">{handle.read()}</span>'
    except Exception:  # pragma: no cover - the icon is decorative only
        pass
    log.warning("services: icon '%s' not found", icon)
    return '<span class="svc-icon"></span>'


def _tag_html(tag: dict[str, str]) -> str:
    """Render one clickable tag.

    A real `<button>` (not a span) so it is reachable by keyboard and exposed
    to screen readers; the JS reads `data-tag-kind`/`data-tag-value`.
    """
    return (
        f'<button type="button" class="svc-tag svc-tag--{tag["kind"]}" '
        f'data-tag-kind="{tag["kind"]}" data-tag-value="{html.escape(tag["value"], quote=True)}">'
        f'{html.escape(tag["label"])}</button>'
    )


def _value_html(value: str) -> str:
    """Escape a value, marking up the "missing information" placeholder."""
    if value == MISSING:
        return f'<span class="svc-missing">{MISSING}</span>'
    return html.escape(value)


def _attr(value: str) -> str:
    """Escape a string for use inside a double-quoted HTML attribute."""
    return html.escape(value, quote=True)


def _card_html(service: dict[str, Any]) -> str:
    """Render one grid card.

    The `data-provider` / `data-group` / `data-type` attributes hold slugs that
    are compared against the dropdown values, and `data-search` holds the
    pre-computed search blob. All filtering reads these attributes, so the card
    markup is the single source of truth for what is filterable.
    """
    tags = "".join(_tag_html(tag) for tag in service["tags"])
    tag_values = " ".join(tag["value"] for tag in service["tags"])
    return f"""<article class="svc-card" id="svc-card-{service['id']}"
  data-id="{service['id']}"
  data-provider="{_attr(_slug(service['provider']))}"
  data-group="{_attr(_slug(service['group']))}"
  data-type="{_attr(_slug(service['type']))}"
  data-tags="{_attr(tag_values)}"
  data-search="{_attr(service['search'])}">
  <button type="button" class="svc-card__open" data-open="{service['id']}">
    <span class="svc-card__icon">{_icon_html(service['icon'])}</span>
    <span class="svc-card__title">{html.escape(service['name'])}</span>
    <span class="svc-card__provider">{_value_html(service['provider'])}</span>
    <span class="svc-card__summary">{_value_html(service['summary'])}</span>
  </button>
  <div class="svc-card__tags">{tags}</div>
</article>"""


#: All user-facing strings, keyed by language. Kept here rather than in the
#: markdown pages so the two languages cannot drift apart structurally.
LABELS: dict[str, dict[str, str]] = {
    "sv": {
        "login": " (kräver inloggning)",
        "to_service": "Till tjänsten",
        "about": "Om tjänsten",
        "close": "Stäng",
        "empty": "Inga tjänster är inlagda ännu. Lägg en markdownfil i mappen "
                 "<code>services/</code> så visas den här.",
        "search": "Sök tjänst, leverantör eller tagg…",
        "search_label": "Sök",
        "provider": "Alla leverantörer",
        "group": "Alla funktionsgrupper",
        "type": "Alla tjänstetyper",
        "reset": "Rensa",
        "noresults": "Inga tjänster matchar filtret.",
        "of": "av",
        "items": "tjänster",
    },
    "en": {
        "login": " (sign-in required)",
        "to_service": "To the service",
        "about": "About the service",
        "close": "Close",
        "empty": "No services are registered yet. Add a markdown file in the "
                 "<code>services/</code> folder and it will appear here.",
        "search": "Search service, provider or tag…",
        "search_label": "Search",
        "provider": "All providers",
        "group": "All function groups",
        "type": "All service types",
        "reset": "Clear",
        "noresults": "No services match the filter.",
        "of": "of",
        "items": "services",
    },
}

DEFAULT_LANG = "en"
SWEDISH_PREFIX = "sv/"


def _modal_sections_html(service: dict[str, Any]) -> str:
    """Render the expandable Access / Guides / About sections of a modal."""
    return "".join(
        f'<details class="svc-section"><summary>{html.escape(section["title"])}</summary>'
        f'<div class="svc-section__body">{section["html"]}</div></details>'
        for section in service["sections"]
    )


def _modal_actions_html(service: dict[str, Any], labels: dict[str, str]) -> str:
    """Render the modal's action buttons.

    "About the service" always exists (it is the generated service page); the
    external link is optional and is annotated when sign-in is required, so a
    researcher knows before clicking.
    """
    external = ""
    if service["link"]:
        login_note = labels["login"] if service["link_login"] else ""
        external = (
            f'<a class="svc-btn svc-btn--primary" href="{_attr(service["link"])}">'
            f'{labels["to_service"]}{login_note}</a>'
        )
    return (
        f'<a class="svc-btn" href="{service["page"]}">{labels["about"]}</a>\n      {external}'
    )


def _modal_html(service: dict[str, Any], labels: dict[str, str]) -> str:
    """Render the modal window for one service.

    Modals are rendered for every service up front (hidden) rather than built
    on demand: the content is static, and pre-rendering keeps the JavaScript
    free of any templating.
    """
    tags = "".join(_tag_html(tag) for tag in service["tags"])
    return f"""<div class="svc-modal" id="svc-modal-{service['id']}" role="dialog" aria-modal="true"
  aria-labelledby="svc-modal-title-{service['id']}" hidden>
  <div class="svc-modal__backdrop" data-close="{service['id']}"></div>
  <div class="svc-modal__panel">
    <button type="button" class="svc-modal__close" data-close="{service['id']}" aria-label="{labels['close']}">&times;</button>
    <p class="svc-modal__icon">{_icon_html(service['icon'])}</p>
    <h2 class="svc-modal__title" id="svc-modal-title-{service['id']}">{html.escape(service['name'])}</h2>
    <p class="svc-modal__provider">{_value_html(service['provider'])}</p>
    <p class="svc-modal__summary">{_value_html(service['summary'])}</p>
    <div class="svc-modal__tags">{tags}</div>
    <div class="svc-modal__sections">{_modal_sections_html(service)}</div>
    <div class="svc-modal__actions">
      {_modal_actions_html(service, labels)}
    </div>
  </div>
</div>"""


def _options(values: Iterable[str]) -> str:
    """Render dropdown `<option>`s, de-duplicated by slug and sorted by label.

    The dropdowns are derived from the service files themselves — adding a new
    provider or function group in a markdown file is all it takes for it to
    appear as a filter. MISSING is skipped: "uppgift saknas" is not a filter.
    """
    seen: dict[str, str] = {}
    for label in values:
        if label and label != MISSING:
            seen.setdefault(_slug(label), label)
    return "".join(
        f'<option value="{_attr(value)}">{html.escape(label)}</option>'
        for value, label in sorted(seen.items(), key=lambda item: item[1].lower())
    )


def _filters_html(services: list[dict[str, Any]], labels: dict[str, str]) -> str:
    """Render the search field, the three dropdowns and the clear button.

    `onsubmit="return false"` keeps Enter in the search field from reloading
    the page — filtering is live on input.
    """
    providers = _options(service["provider"] for service in services)
    groups = _options(service["group"] for service in services)
    types = _options(service["type"] for service in services)
    return f"""<form class="svc-filters" role="search" onsubmit="return false;">
    <input type="search" class="svc-filters__search" name="q" placeholder="{labels['search']}" aria-label="{labels['search_label']}">
    <select name="provider" aria-label="{labels['provider']}"><option value="">{labels['provider']}</option>{providers}</select>
    <select name="group" aria-label="{labels['group']}"><option value="">{labels['group']}</option>{groups}</select>
    <select name="type" aria-label="{labels['type']}"><option value="">{labels['type']}</option>{types}</select>
    <button type="button" class="svc-filters__reset" disabled>{labels['reset']}</button>
  </form>"""


def _archive_html(services: list[dict[str, Any]], labels: dict[str, str]) -> str:
    """Render the complete archive block that replaces the marker."""
    if not services:
        return f'<p class="svc-empty">{labels["empty"]}</p>'

    cards = "".join(_card_html(service) for service in services)
    modals = "".join(_modal_html(service, labels) for service in services)

    return f"""<div class="svc-archive" data-count="{len(services)}"
  data-label-of="{labels['of']}" data-label-items="{labels['items']}">
  {_filters_html(services, labels)}
  <p class="svc-count" aria-live="polite"></p>
  <div class="svc-grid">{cards}</div>
  <p class="svc-noresults" hidden>{labels['noresults']}</p>
  <div class="svc-modals">{modals}</div>
</div>"""


# --------------------------------------------------------------------------- #
# MkDocs entry point
# --------------------------------------------------------------------------- #


def on_page_markdown(markdown: str, page: Any, config: Any, files: Any, **kwargs: Any) -> str:
    """MkDocs hook: replace the archive marker with the generated catalogue.

    Args:
        markdown: The page source. Returned unchanged unless it holds MARKER.
        page: The MkDocs page; its `file.src_path` gives both the language and
            the folder to scan for service files.
        config: The MkDocs config, used for `docs_dir`.
        files: Unused; part of the MkDocs hook signature.

    Returns:
        The page markdown, with the marker replaced by raw HTML.
    """
    if MARKER not in markdown:
        return markdown
    src = page.file.src_path.replace(os.sep, "/")
    lang = "sv" if src.startswith(SWEDISH_PREFIX) else DEFAULT_LANG
    services = _collect(page, config)
    return markdown.replace(MARKER, _archive_html(services, LABELS[lang]))

File: docs/javascripts/service_catalog/services.js
/* ==========================================================================
   Digital service catalogue — search, filtering, tags and modals
   --------------------------------------------------------------------------
   WHY THIS EXISTS
   The site is a static MkDocs build with no backend, so all filtering happens
   in the browser against markup generated at build time by
   service_catalog/hooks/services.py. Every card carries the data it is
   filtered on (`data-search`, `data-provider`, `data-group`, `data-type`),
   which keeps this file free of any knowledge about individual services.

   INPUTS   The generated `.svc-archive` block, plus the query string.
   OUTPUT   Cards shown/hidden, a result counter, open/closed modals and a
            shareable URL (`?q=&provider=&group=&type=&service=`).

   HIDING CONTRACT — READ BEFORE CHANGING
   A card is hidden by setting its `hidden` attribute. That only takes effect
   because services.css restates `.svc-card[hidden] { display: none; }`; the
   browser's default `[hidden]` rule loses to the `.svc-card { display: flex }`
   rule. The original bug was exactly this: the counter updated but every card
   stayed on screen. If you hide a new element here, give it the same rule.

   DEPENDENCIES
   None. Written in ES5-compatible vanilla JS and re-initialised on Material's
   `document$` observable, because instant navigation swaps the page body
   without a full reload.
   ========================================================================== */
(function () {
  "use strict";

  /* Prefix of the generated modal element ids (`svc-modal-<service id>`). */
  var MODAL_ID_PREFIX = "svc-modal-";
  /* Marks an archive that has already been wired up, so instant navigation
     cannot attach a second set of listeners to the same DOM. */
  var READY_FLAG = "1";
  /* Fallbacks used only if the generated labels are missing (Swedish page). */
  var FALLBACK_LABEL_OF = "av";
  var FALLBACK_LABEL_ITEMS = "tjänster";

  function init() {
    var archive = document.querySelector(".svc-archive");
    if (!archive || archive.dataset.ready === READY_FLAG) return;
    archive.dataset.ready = READY_FLAG;

    var form = archive.querySelector(".svc-filters");
    var search = form.querySelector('input[name="q"]');

    /* Keys match both the `<select name>` and the card `data-*` attribute,
       which is what lets `matches()` stay generic. Adding a fourth filter is
       a matter of generating the select + data attribute and adding it here. */
    var selects = {
      provider: form.querySelector('select[name="provider"]'),
      group: form.querySelector('select[name="group"]'),
      type: form.querySelector('select[name="type"]')
    };

    var reset = form.querySelector(".svc-filters__reset");
    var cards = Array.prototype.slice.call(archive.querySelectorAll(".svc-card"));
    var count = archive.querySelector(".svc-count");
    var noresults = archive.querySelector(".svc-noresults");
    /* Element that had focus before a modal opened, so focus can be restored
       when it closes (keyboard and screen-reader users would otherwise be
       dropped back at the top of the document). */
    var lastFocus = null;

    /**
     * Decide whether a card passes the current search text and dropdowns.
     * Free text matches the pre-computed `data-search` blob (name, provider,
     * group, type, summary, access and all tag labels, lowercased at build
     * time). Dropdowns compare slugs, so they are accent- and case-safe.
     * @param {HTMLElement} card
     * @returns {boolean} true when the card should stay visible
     */
    function matches(card) {
      var q = search.value.trim().toLowerCase();
      if (q) {
        var terms = q.split(/\s+/);
        for (var i = 0; i < terms.length; i++) {
          if (card.dataset.search.indexOf(terms[i]) === -1) return false;
        }
      }
      for (var key in selects) {
        var value = selects[key].value;
        if (value && card.dataset[key] !== value) return false;
      }
      return true;
    }

    /**
     * Apply the current filter state to the grid.
     * Side effects: toggles `hidden` on every card, updates the counter and
     * the "no results" message, enables/disables Clear, and (unless
     * `pushState === false`) rewrites the query string.
     * @param {boolean} [pushState] pass false during initial load
     */
    function apply(pushState) {
      var visible = 0;
      cards.forEach(function (card) {
        var show = matches(card);
        /* See the HIDING CONTRACT note at the top of this file. */
        card.hidden = !show;
        if (show) visible++;
      });

      count.textContent = visible + " " + (archive.dataset.labelOf || FALLBACK_LABEL_OF) +
        " " + cards.length + " " + (archive.dataset.labelItems || FALLBACK_LABEL_ITEMS);

      noresults.hidden = visible !== 0;

      var dirty = !!search.value || !!selects.provider.value ||
        !!selects.group.value || !!selects.type.value;
      reset.disabled = !dirty;

      if (pushState !== false) writeUrl();
    }

    /**
     * Mirror the current state into the query string with replaceState, so a
     * filtered view (and an open card) can be copied, shared and reloaded
     * without filling the browser history with one entry per keystroke.
     * @param {string} [openId] service id of the modal to record; omit to
     *        reuse whichever modal is currently open
     */
    function writeUrl(openId) {
      var params = new URLSearchParams();
      if (search.value) params.set("q", search.value);
      for (var key in selects) if (selects[key].value) params.set(key, selects[key].value);
      var open = typeof openId !== "undefined" ? openId : currentOpenId();
      if (open) params.set("service", open);
      var qs = params.toString();
      history.replaceState(null, "", qs ? "?" + qs : location.pathname);
    }

    /** @returns {string} id of the open service modal, or "" if none is open */
    function currentOpenId() {
      var open = archive.querySelector(".svc-modal:not([hidden])");
      return open ? open.id.replace(MODAL_ID_PREFIX, "") : "";
    }

    /* ---- modal ----------------------------------------------------------
       Modals are pre-rendered and hidden; opening one never touches the grid,
       so closing it returns the reader to exactly the filtered view they came
       from. */

    /**
     * Open the modal for a service.
     * @param {string} id service id (the markdown filename, slugified)
     * @param {HTMLElement} [focusEl] element to return focus to on close
     */
    function openModal(id, focusEl) {
      var modal = document.getElementById(MODAL_ID_PREFIX + id);
      if (!modal) return;
      closeModal(true); /* only one modal at a time */
      lastFocus = focusEl || document.activeElement;
      modal.hidden = false;
      document.body.classList.add("svc-modal-open");
      var close = modal.querySelector(".svc-modal__close");
      if (close) close.focus();
      writeUrl(id);
    }

    /**
     * Close the open modal.
     * @param {boolean} [silent] when true, skip focus restore and URL update
     *        (used when immediately opening another modal or filtering)
     */
    function closeModal(silent) {
      var open = archive.querySelector(".svc-modal:not([hidden])");
      if (open) open.hidden = true;
      document.body.classList.remove("svc-modal-open");
      if (!silent) {
        if (lastFocus && lastFocus.focus) lastFocus.focus();
        writeUrl("");
      }
    }

    /* ---- event handling -------------------------------------------------
       One delegated click listener on the archive covers cards, close buttons
       and tags. Delegation matters because every card, tag and modal is
       generated markup: no per-element wiring has to be kept in sync. */

    archive.addEventListener("click", function (event) {
      var open = event.target.closest("[data-open]");
      if (open) {
        event.preventDefault();
        openModal(open.dataset.open, open);
        return;
      }

      var close = event.target.closest("[data-close]");
      if (close) {
        event.preventDefault();
        closeModal();
        return;
      }

      var tag = event.target.closest(".svc-tag");
      if (tag) {
        event.preventDefault();
        var kind = tag.dataset.tagKind;
        var value = tag.dataset.tagValue;
        if (selects[kind]) {
          /* Clicking the active tag again clears the filter (toggle). */
          selects[kind].value = selects[kind].value === value ? "" : value;
          search.value = "";
        } else {
          /* Tag families without a dropdown (data, location) fall back to a
             free-text search on the label. */
          search.value = search.value === tag.textContent.trim() ? "" : tag.textContent.trim();
        }
        /* Silent close: the reader asked for a new filtered view, so the URL
           should describe the filter, not the card they came from. */
        closeModal(true);
        document.body.classList.remove("svc-modal-open");
        apply();
        archive.scrollIntoView({ behavior: "smooth", block: "start" });
      }
    });

    document.addEventListener("keydown", function (event) {
      if (event.key === "Escape" && archive.querySelector(".svc-modal:not([hidden])")) {
        closeModal();
      }
    });

    search.addEventListener("input", function () { apply(); });
    Object.keys(selects).forEach(function (key) {
      selects[key].addEventListener("change", function () { apply(); });
    });
    reset.addEventListener("click", function () {
      search.value = "";
      Object.keys(selects).forEach(function (key) { selects[key].value = ""; });
      apply();
    });

    /* ---- restore state from the URL ------------------------------------ */

    var params = new URLSearchParams(location.search);
    search.value = params.get("q") || "";
    Object.keys(selects).forEach(function (key) {
      var value = params.get(key);
      if (value) selects[key].value = value;
    });
    apply(false); /* false: don't rewrite the URL we just read */
    if (params.get("service")) openModal(params.get("service"));
  }

  /* Material for MkDocs uses instant navigation: the body is replaced without
     a page load, so `DOMContentLoaded` fires only once. `document$` emits on
     every page view; the READY_FLAG guard keeps that idempotent. */
  if (window.document$ && typeof window.document$.subscribe === "function") {
    window.document$.subscribe(init);
  } else {
    document.addEventListener("DOMContentLoaded", init);
  }
})();
